web/src/echo.ts

```diff
@@ -10,27 +12,103 @@ import * as markdown from "./markdown";
 import * as message_lists from "./message_lists";
 import * as message_live_update from "./message_live_update";
 import * as message_store from "./message_store";
+import type {DisplayRecipientUser, Message} from "./message_store";
 import * as message_util from "./message_util";
 import * as people from "./people";
 import * as pm_list from "./pm_list";
 import * as recent_view_data from "./recent_view_data";
 import * as rows from "./rows";
 import * as sent_messages from "./sent_messages";
+import {server_message_schema} from "./server_message";
 import {current_user} from "./state_data";
 import * as stream_data from "./stream_data";
 import * as stream_list from "./stream_list";
 import * as stream_topic_history from "./stream_topic_history";
+import type {TopicLink} from "./types";
 import * as util from "./util";
 
 // Docs: https://zulip.readthedocs.io/en/latest/subsystems/sending-messages.html
 
-const waiting_for_id = new Map();
-let waiting_for_ack = new Map();
+const server_message_schema_with_flags = server_message_schema.and(
+    z.object({
+        local_id: z.string().optional(),
+        flags: z.array(z.string()),
+    }),
+);
+
+type ServerMessage = z.output<typeof server_message_schema_with_flags>;
```

Comment:
I think we can use `type RawMessage` instead of creating a new type `ServerMessage`. 
I used a similar schema in this commit ae05786a24c8ffd0fa6ac38ba0fc4b3a2bdd46ca to infer the type `RawMessage`.

I believe we should avoid making too many new types since it will create a problem migrating other files due to these different types with barely 1 key different from the other.

Reply:
I mostly agree with your point except for few things
- I think messages recieved from server should have an independent type as `raw_message` and `server_message` differ in couple of fields.
- Also even if we do not create a new type and use `Omit<RawMessage, "foo field"> & {bar fields}`, I think there are few incomplete type fields in `RawMessage` that have to be fixed first.